[SelectionDAG] Fix null pointer dereference in resolveDanglingDebugInfo

MetalOxideSemi opened the pull request:
## Summary
Fix null pointer dereference in `SelectionDAGBuilder::resolveDanglingDebugInfo`.

## Problem
`Val.getNode()->getIROrder()` is called before checking if `Val.getNode()` is null, causing crashes when compiling code with debug info that contains aggregate constants with nested empty structs.

## Solution
Move the `ValSDNodeOrder` declaration inside the `if (Val.getNode())` block.

## Test Case
Reproduces with aggregate types containing nested empty structs:
```llvm
%3 = insertvalue { { i1, {} }, ptr, { { {} }, { {} } }, i64 } 
     { { i1, {} } zeroinitializer, ptr null, { { {} }, { {} } } zeroinitializer, i64 2 }, 
     ptr %2, 1, !dbg !893

## Crash stack
0.      Program arguments: llc-20 -O3 -mcpu=native -relocation-model=pic -filetype=obj /cloudide/workspace/temp/sf.ll -o /dev/null
1.      Running pass 'Function Pass Manager' on module '/cloudide/workspace/temp/sf.ll'.
2.      Running pass 'X86 DAG->DAG Instruction Selection' on function '@filter_create'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libLLVM.so.20.1 0x00007ff87ebbdf86 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 54
1  libLLVM.so.20.1 0x00007ff87ebbbb90 llvm::sys::RunSignalHandlers() + 80
2  libLLVM.so.20.1 0x00007ff87ebbe640
3  libpthread.so.0 0x00007ff87db79140
4  libLLVM.so.20.1 0x00007ff87f3fd2ff llvm::SelectionDAGBuilder::resolveDanglingDebugInfo(llvm::Value const*, llvm::SDValue) + 303
5  libLLVM.so.20.1 0x00007ff87f3fda5e llvm::SelectionDAGBuilder::getValue(llvm::Value const*) + 142
6  libLLVM.so.20.1 0x00007ff87f3fe79f llvm::SelectionDAGBuilder::getValueImpl(llvm::Value const*) + 3343
7  libLLVM.so.20.1 0x00007ff87f3fda34 llvm::SelectionDAGBuilder::getValue(llvm::Value const*) + 100
8  libLLVM.so.20.1 0x00007ff87f3fc1ab llvm::SelectionDAGBuilder::visitInsertValue(llvm::InsertValueInst const&) + 603
9  libLLVM.so.20.1 0x00007ff87f3eeaf7 llvm::SelectionDAGBuilder::visit(llvm::Instruction const&) + 327
10 libLLVM.so.20.1 0x00007ff87f4904b8 llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void, true, llvm::BasicBlock>, false, true>, llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void, true, llvm::BasicBlock>, false, true>, bool&) + 72
11 libLLVM.so.20.1 0x00007ff87f490304 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) + 5956
12 libLLVM.so.20.1 0x00007ff87f48e2b4 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) + 372
13 libLLVM.so.20.1 0x00007ff87f48c689 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) + 169
14 libLLVM.so.20.1 0x00007ff87efb8e32 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 610
15 libLLVM.so.20.1 0x00007ff87ed104be llvm::FPPassManager::runOnFunction(llvm::Function&) + 638
16 libLLVM.so.20.1 0x00007ff87ed15ff3 llvm::FPPassManager::runOnModule(llvm::Module&) + 51
17 libLLVM.so.20.1 0x00007ff87ed10c11 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 1105
18 llc-20          0x000055972ce77dc1 main + 9649
19 libc.so.6       0x00007ff87d68ad7a __libc_start_main + 234
20 llc-20          0x000055972ce7247a _start + 42
```

## Testing

Added regression tests in:
- `CodeGen/X86/selectiondag-dbgvalue-null-crash.ll`
- `CodeGen/AArch64/selectiondag-dbgvalue-null-crash.ll`

**Note:** Tests are currently restricted to x86-64 Linux hosts only, as they appear to expose deeper issues in DWARF generation on certain host/target combinations that require further investigation.

## Related PRs

This supersedes:
- #173500 - Initial fix, reverted due to test failures on Darwin and other platforms
- #173836 - Second attempt with `UNSUPPORTED: system-darwin`, still failed on some targets

MetalOxideSemi committed:
[SelectionDAG] Fix null pointer dereference in resolveDanglingDebugInfo

When handling dbg_value intrinsics for aggregate types containing
empty structs, Val.getNode() could return null, causing a crash
when attempting to dereference it.

This patch adds a null check before accessing Val.getNode() in
resolveDanglingDebugInfo, preventing the crash while maintaining
correct debug info handling for valid cases.

The issue was triggered by:
1. A dbg_value intrinsic referencing an aggregate type with empty structs
2. An insertvalue operation that gets lowered to a null SDNode
3. Attempting to dereference the null node in resolveDanglingDebugInfo

Added regression test in CodeGen/{X86|AArch64}/selectiondag-dbgvalue-null-crash.ll

arsenm commented on llvm/test/CodeGen/AArch64/selectiondag-dbgvalue-null-crash.ll:
```diff
@@ -0,0 +1,45 @@
+; RUN: llc < %s -mtriple=aarch64-unknown-linux-gnu | FileCheck %s
+; REQUIRES: system-linux
+; REQUIRES: x86-host
+; REQUIRES: x86-registered-target, aarch64-registered-target
```
None of these requires are required. None of this is host dependent, the failure was due to the default target triple

arsenm commented on llvm/test/CodeGen/X86/selectiondag-dbgvalue-null-crash.ll:
```diff
@@ -0,0 +1,42 @@
+;
+; FIXME: This test appears to reveal deeper bugs in DWARF generation logic
+; under certain host/target combinations. For now, we restrict it to Linux
+; hosts only until those underlying issues are fixed.
```
Comment is out of date?

MetalOxideSemi commented:
@arsenm Hi, I managed to borrow a macOS device and it turned out to have nothing to do with host system. Targeting apple-darwin will always crash no matter which host it is running on, which matches your observation on the failure of the previous PR is due to using default target (hence running on a macOS device causes it to targeting macOS and crash). I've fixed llvm-lit instruction and added XFAIL on darwin targets. Thanks for your patience!
